### PR TITLE
Navigation animations

### PR DESCRIPTION
**Pull request by PixelPirate**
* **Please check if the PR fulfills these requirements**
- [x] The commit message makes sense
- [x] The changes have been tested locally
- [x] Are there any breaking changes

* **What kind of change does this PR introduce?**
(Bug fix, feature, docs update, ...)
This PR adds animations when entering into or exiting out of a sub menu. In addition to #608, this helps with navigating the menus, especially when you haven't memorized all the menus.
Animation

* **What is the current behavior?**
(You can also link to an open issue here)
Sub menus are displayed instantly when opened.

* **What is the new behavior (if this is a feature change)?**
Going deeper into the menu hierarchy shows a right to left animation. Going back displays a left to right animation.

* **Other information**:
During the animation, a second buffer for the display contents is created. I didn't see any problems with that on my iron, but I don't know what the memory constraints are.
If you want the animations to be optional, I can add a menu entry for it.

### CODE REVIEW

File: workspace/TS100/Core/Src/gui.cpp [hunk 1]
```diff
@@ -931,7 +963,7 @@ void gui_Menu(const menuitem *menu) {
 			osDelay(40);
 			lcdRefresh = false;
 		}
-		if ((xTaskGetTickCount() - lastButtonTime) > (1000 * 30)) {
```
Comment (Ralim) [inline workspace/TS100/Core/Src/gui.cpp:934]: Good catch

**Comment by Ralim** [comment]
A+ Implementation, Love it :)
I'm curious if this would be simpler with using the OLED's build in scrolling - but also not sure if that works sideways. Either way, 1000% happy with this.

My only concern here is I would rather not allocate these buffers on the stack but instead statically allocate at compile time (we have the room), as I tend to keep the stacks with only a smallish amount of surplus. It also helps prevent mystery bugs as compiling will just fail instead.

Could you by chance do a small refactor to allocate buffers statically at compile time ?
Don't care if you make it static inside the OLED class (ie have it hold two) or if you make it static in `gui.cpp`

**Comment by PixelPirate** [comment]
Thank you, and I will change the buffer. Having a static inside the OLED class was actually how I did it on my first attempt :)

**Comment by Ralim** [comment]
Ah bugger, could have gone with less work slightly :D 
Another small benefit of putting in the OLED class is it keeps it nicely organised when looking at ram allocation breakdown tables (for when we run low again) :D 
Ping me when you have time to update it :D

**Comment by PixelPirate** [comment]
@Ralim I updated the PR :)

File: workspace/TS100/Core/Src/OLED.cpp [hunk 2]
```diff
@@ -62,6 +63,14 @@ uint8_t OLED_Setup_Array[] = {
 const uint8_t REFRESH_COMMANDS[17] = { 0x80, 0xAF, 0x80, 0x21, 0x80, 0x20, 0x80,
 		0x7F, 0x80, 0xC0, 0x80, 0x22, 0x80, 0x00, 0x80, 0x01, 0x40 };
 
+static uint8_t easeInOutTiming(uint8_t t) {
```
Comment (Ralim) [inline workspace/TS100/Core/Src/OLED.cpp:66]: Could you add a comment on what these two are for rather than magic math equations?

**Comment by Ralim** [comment]
@PixelPirate Tiny change :D Then I'm good to merge it.

File: workspace/TS100/Core/Src/OLED.cpp (hunk 2 above)
Comment (PixelPirate) [reply workspace/TS100/Core/Src/OLED.cpp:73]: Sure, I added some documentation :)

**Pull request merged**